### PR TITLE
Improvement for SI-2251, failure to lub f-bounds.

### PR DESCRIPTION
**Pull request by paulp**
After a great struggle, I realized that the major reason that
code like this still doesn't compile:

  List(Stream(), List())

is that we were poisoning the computed lub in mergePrefixAndArgs by
throwing in Any when the max recursion depth was reached. I modified
it to return NoType instead, which allowed me to teach lublist to
recognize what has happened and fall back to a weaker type, one
which does not contain recursive bounds. This enables the lubbing
process to complete. The most elusive lub, defeated. Notice also
that the refinement members are correctly parameterized on Nothing,
rather than on Any as has often been the case.

(Re: Any/Nothing, see the change to test/files/run/lub-visibility.check.)

scala> List(Stream(), List())
res0: List[scala.collection.immutable.LinearSeq[Nothing] with scala.collection.AbstractSeq[Nothing]{def companion: scala.collection.generic.GenericCompanion[scala.collection.immutable.LinearSeq with scala.collection.AbstractSeq]; def reverse: scala.collection.immutable.LinearSeq[Nothing] with scala.collection.AbstractSeq[Nothing]{def companion: scala.collection.generic.GenericCompanion[scala.collection.immutable.LinearSeq with scala.collection.AbstractSeq]; def reverse: scala.collection.immutable.LinearSeq[Nothing] with scala.collection.AbstractSeq[Nothing]{def reverse: scala.collection.immutable.LinearSeq[Nothing] with scala.collection.AbstractSeq[Nothing]; def dropRight(n: Int): scala.collection.immutable.LinearSeq[Nothing] with scala.collection.AbstractSeq[Nothing]; def takeRight(n: ...

### CODE REVIEW

**Comment by paulp** [comment]
Review by @odersky or @adriaanm .

**Comment by jsuereth** [comment]
If you want, we can skype a call here and you can describe what you did.  Don't really think I should be reviewing this code normally though :)

**Comment by paulp** [comment]
We're supposed to skype anyway, right? We can throw this in.  If you're still around today.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -6469,13 +6499,13 @@ trait Types extends api.Types { self: SymbolTable =>
         }
     }
     def lub1(ts0: List[Type]): Type = {
-      val (ts, tparams) = stripExistentialsAndTypeVars(ts0)
+      val (ts, tparams)            = stripExistentialsAndTypeVars(ts0)
```
Comment (jsuereth) [inline src/reflect/scala/reflect/internal/Types.scala:6502]: These refactored lines are immediate indicator of author to me.  Well done, sir.

**Comment by jsuereth** [comment]
Based on our phone call and my hazy understanding of typer, this looks like a great addition.   I totally see the magic in https://github.com/scala/scala/pull/1122/files#L0R6258 and the effects appear correct.  I'd be up for merging if @odersky or @adriaanm don't get a chance to review soon.  Tests pass, general improvement.  Thanks again!

**Comment by paulp** [comment]
That review apparently will have to suffice for now.  I'm sure we'll revisit the specifics.

**Pull request merged**